ui/src/views/ThirdPartyLogin/index.jsx

```diff
@@ -0,0 +1,111 @@
+import { hot } from 'react-hot-loader';
+import React, { Component } from 'react';
+import { graphql } from 'react-apollo';
+import { scopeIntersection } from 'taskcluster-lib-scopes';
+import { parse } from 'qs';
+import Spinner from '@mozilla-frontend-infra/components/Spinner';
+import Homepage from '../../components/Homepage';
+import AuthConsent from '../../components/AuthConsent';
+import Dashboard from '../../components/Dashboard';
+import { withAuth } from '../../utils/Auth';
+import fromNowJSON from '../../utils/fromNowJSON';
+import thirdPartyLoginQuery from './thirdPartyLogin.graphql';
+
+@hot(module)
+@withAuth
+@graphql(thirdPartyLoginQuery, {
+  skip: ({ user }) => !user,
+  options: () => ({
+    fetchPolicy: 'network-only',
+  }),
+})
+export default class ThirdPartyLogin extends Component {
+  query = this.props.location.search.slice(1);
+
+  parsedQuery = parse(this.query);
+
+  state = {
+    formData: null,
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    const {
+      data,
+      user,
+      location: { search },
+    } = props;
+    const query = parse(search.slice(1));
+    const scopes = query.scope.split(' ');
+    const registeredClientId = query.client_id;
+
+    if (
+      !data ||
+      state.formData ||
+      !(data.currentScopes instanceof Array) ||
+      !query.transactionID
+    ) {
+      return null;
+    }
+
+    return {
+      formData: {
+        ...state.formData,
+        scopes: scopeIntersection(scopes, data.currentScopes),
```

Comment:
As a note for future reviewers, this is not the authoritative place that this scope-intersection operation occurs: `scopeIntersection` is also called in `services/web-server/src/servers/oauth2.js` and then enforced by the use of `authorizedScopes` in the `auth.createClient` call in that file.